tests/components/filter/test_sensor.py

```diff
@@ -15,309 +16,307 @@
+                ha.State("sensor.test_monitored", 19.0, last_changed=t_2),
+                ha.State("sensor.test_monitored", 18.2, last_changed=t_3),
+            ]
         }
 
-        with assert_setup_component(1, "sensor"):
-            assert setup_component(self.hass, "sensor", config)
-            self.hass.block_till_done()
+    with patch(
+        "homeassistant.components.history.state_changes_during_period",
+        return_value=fake_states,
+    ):
+        with patch(
+            "homeassistant.components.history.get_last_state_changes",
+            return_value=fake_states,
+        ):
+            with assert_setup_component(1, "sensor"):
+                assert await async_setup_component(hass, "sensor", config)
+                await hass.async_block_till_done()
 
-            for value in self.values:
-                self.hass.states.set(config["sensor"]["entity_id"], value.state)
-                self.hass.block_till_done()
+            for value in values:
+                hass.states.async_set(config["sensor"]["entity_id"], value.state)
+                await hass.async_block_till_done()
 
-            state = self.hass.states.get("sensor.test")
-            assert "18.05" == state.state
+            state = hass.states.get("sensor.test")
+            if missing:
+                assert "18.05" == state.state
+            else:
+                assert "17.05" == state.state
 
-    def test_chain_history(self, missing=False):
-        """Test if filter chaining works."""
-        self.init_recorder()
-        config = {
-            "history": {},
-            "sensor": {
-                "platform": "filter",
-                "name": "test",
-                "entity_id": "sensor.test_monitored",
-                "filters": [
-                    {"filter": "outlier", "window_size": 10, "radius": 4.0},
-                    {"filter": "lowpass", "time_constant": 10, "precision": 2},
-                    {"filter": "throttle", "window_size": 1},
-                ],
-            },
-        }
-        t_0 = dt_util.utcnow() - timedelta(minutes=1)
-        t_1 = dt_util.utcnow() - timedelta(minutes=2)
-        t_2 = dt_util.utcnow() - timedelta(minutes=3)
-        t_3 = dt_util.utcnow() - timedelta(minutes=4)
 
-        if missing:
-            fake_states = {}
-        else:
-            fake_states = {
-                "sensor.test_monitored": [
-                    ha.State("sensor.test_monitored", 18.0, last_changed=t_0),
-                    ha.State("sensor.test_monitored", "unknown", last_changed=t_1),
-                    ha.State("sensor.test_monitored", 19.0, last_changed=t_2),
-                    ha.State("sensor.test_monitored", 18.2, last_changed=t_3),
-                ]
-            }
+async def test_chain_history_missing(hass, values):
+    """Test if filter chaining works when recorder is enabled but the source is not recorded."""
+    await test_chain_history(hass, values, missing=True)
```

Comment:
It's a bit weird that we are calling the test function `test_chain_history` from this test. Usually we separate helpers from test functions.